src/lib/core/datetime/date-adapter.ts

```diff
@@ -0,0 +1,172 @@
+   * @param locale The locale to use.
+   * @returns The first day of the week (0-indexed, 0 = Sunday).
+   */
+  abstract getFirstDayOfWeek(locale?: any): number;
+
+  /**
+   * Creates a date with the given year, month, and date.
+   * @param year The full year of the date. (e.g. 89 means the year 89, not the year 1989).
+   * @param month The month of the date (0-indexed, 0 = January). If `month` is less than 0 or
+   *     greater than 11, it should roll into the previous / next year.
+   * @param date The date of month of the date. If `date` is less than 1 or greater than the number
+   *     of days in the `month`, it should roll into the previous / next month.
+   * @returns The new date.
+   */
+  abstract create(year: number, month: number, date: number): D;
+
+  /**
+   * Gets today's date.
+   * @returns Today's date.
+   */
+  abstract today(): D;
+
+  /**
+   * Parses a date from a value.
+   * @param value The value to parse.
+   * @param fmt The format of the value.
+   * @param locale The locale of the value.
+   * @returns The parsed date, or null if date could not be parsed.
+   */
+  abstract parse(value: any, fmt?: any, locale?: any): D | null;
+
+  /**
+   * Formats a date as a string.
+   * @param date The value to parse.
+   * @param fmt The format to use for the result string.
+   * @param locale The locale to use for the result string.
+   * @returns The parsed date, or null if date could not be parsed.
+   */
+  abstract format(date: D, fmt?: any, locale?: any): string;
+
+  /**
+   * Adds the given number of years, months, and days to the given date.
+   * @param date The date to add to.
+   * @param amount The number of years, months, and days to add (may be negative).
+   * @returns A new date equal to the original with the given amount of time added.
+   */
+  abstract add(date: D, amount: {years?: number, months?: number, days?: number}): D;
+
+  /**
+   * Clones the given date.
+   * @param date The date to clone
+   * @returns A new date equal to the given date.
+   */
+  clone(date: D): D {
+    return this.create(this.getYear(date), this.getMonth(date), this.getDate(date));
+  }
+
+  /**
+   * Compares two dates.
+   * @param first The first date to compare.
+   * @param second The second date to compare.
+   * @returns 0 if the dates are equal, a number less than 0 if the first date is earlier,
+   *     a number greater than 0 if the first date is later.
+   */
+  compare(first: D, second: D): number {
+    return this.getYear(first) - this.getYear(second) ||
+        this.getMonth(first) - this.getMonth(second) ||
+        this.getDate(first) - this.getDate(second);
+  }
+
+  /**
+   * Checks if two dates are equal.
+   * @param first The first date to check.
+   * @param second The second date to check.
+   * @returns {boolean} Whether the two dates are equal.
+   *     Null dates are considered equal to other null dates.
+   */
+  equals(first: D | null, second: D | null): boolean {
```

Comment:
Similar comment as above for `equals` and `compare`: you'll want different things for the time variants. Maybe just call them `compareDate` and `sameDate`?